libraries/abstractions/common_io/test/test_scripts/uart/test_iot_uart_test.py

```diff
@@ -0,0 +1,259 @@
+        self._serial.write(cmd.encode("utf-8"))
+        self._serial.write("\r\n".encode("utf-8"))
+
+        dut_res =  str(self._serial.read_until(terminator=serial.to_bytes([ord(c) for c in 'Ignored \n\r'])))
+        dut_result = dut_res_pattern.search(dut_res)
+        print(dut_res)
+        t_shell.join()
+
+        with open(self.rpi_output_file, "r") as f:
+            rpi_output = str(f.read());
+            print(rpi_output)
+            pi_result = pi_regex.search(rpi_output)
+
+        s.send(b'E')
+        s.close()
+        self.clean()
+
+        return "Pass" if pi_result and dut_result else "Fail"
+
+    def test_IotUartBaudChange(self):
+        """
+        Test body of uart baud change test. Return pass if expected messages are recieved
+        :return: 'Pass' or 'Fail'
+        """
+        result_matches = []
+        dut_result = None
+        dut_res_pattern = re.compile(r"AFQP_AssistedIotUARTBaudChange\W\sPASS")
+        pi_res_pattern = re.compile(r"UART\sread-write")
+
+        t_shell = threading.Thread(target=self.run_shell_script,
+                                   args=(" ".join([self.shell_script, self._ip, self._login, self._pwd, '2']),))
+        t_shell.start()
+
+        # IMPORTANT: without this, the socket won't connect on mac; reason not known.
+        sleep(1)
+
+        # Create a tcp type socket with AF_INET address family.
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        time_out = 10
+        # Wait until connection with the process on rpi is established.
+        while s.connect_ex((self._ip, self.port)) != 0 and time_out > 0:
+            time_out -= 1
+            sleep(1)
+        if time_out == 0:
+            print("Socket connection cannot be established")
+            s.close()
+            return "Fail"
+
+        print("connnected successfully")
+
+        self._serial.reset_input_buffer()
+
+        cmd = "iot_tests test 1 2"
+        self._serial.write("\r\n".encode("utf-8"))
+        self._serial.write(cmd.encode("utf-8"))
+        self._serial.write("\r\n".encode("utf-8"))
+
+        dut_res =  str(self._serial.read_until(terminator=serial.to_bytes([ord(c) for c in 'Ignored \n\r'])))
+        dut_result = dut_res_pattern.search(dut_res)
+
+        print(dut_res)
+        t_shell.join()
+
+        with open(self.rpi_output_file, "r") as f:
+            result_matches = pi_res_pattern.findall(str(f.read()))
+        s.send(b'E')
+        s.close()
+        self.clean()
+
+        return "Pass" if len(result_matches) >= 2 and dut_result else "Fail"
+
+    def test_IotUartWriteAsync(self):
+        """
+        Test body of uart write async test. Return pass if expected message is received
+        :return: 'Pass' or 'Fail'
+        """
+        pi_result = None
+        dut_result = None
+        dut_res_pattern = re.compile(r"AFQP_AssistedIotUARTWriteAsync\W\sPASS")
+        pi_res_pattern = re.compile(r"\b(?:\\xaa){198}\b")
```

Comment:
why it is 198 here? Look like dut is sending out 200 bytes.

Reply:
https://github.com/aws/amazon-freertos/blob/2049152156e5ffc7d2d775da29b527487cc857b1/libraries/abstractions/common_io/test/test_iot_uart.c#L418 We lose one byte as the UART assisted tests use a '\n' byte to signal the write is done. In my PR I added a '\n' to the end of the buffer, but this change was lost when reviving that PR. We need to set cpBufferLarge[ testIotUART_BUFFER_LENGTH_LARGE  ] = '\n' and change this  to 199 bytes. I believe this was 198 as I was using '\r\n' but '\n' is sufficient and easier.

Reply:
I will update it to 200